segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeper/BookKeeperLog.java

```diff
@@ -308,23 +336,76 @@ private boolean processWrites(List<Write> toExecute) {
                 return false;
             }
         }
+
         return true;
     }
 
     /**
-     * Gets an ordered list of Writes to execute.
+     * Checks each Write in the given list if it is pointing to a closed WriteLedger. If so, it verifies if the write has
+     * actually been committed (in case we hadn't been able to determine its outcome) and updates the Ledger, if needed.
+     *
+     * @param writes An ordered list of Writes to inspect and update.
+     * @return True if any of the Writes in the given list has been modified (either completed or had its WriteLedger
+     * changed).
      */
-    private List<Write> getWritesToExecute() {
-        // Clean up the write queue of all finished writes that are complete (successfully or failed for good)
-        if (!this.writes.removeFinishedWrites()) {
-            return Collections.emptyList();
+    private boolean handleClosedLedgers(List<Write> writes) {
+        if (writes.size() == 0 || !writes.get(0).getWriteLedger().ledger.isClosed()) {
+            // Nothing to do. We only need to check the first write since, if a Write failed with LedgerClosed, then the
+            // first write must have failed for that reason (a Ledger is closed implies all ledgers before it are closed too).
+            return false;
         }
 
-        // Calculate how much estimated space there is in the current ledger.
-        final long maxTotalSize = this.config.getBkLedgerMaxSize() - getWriteLedger().ledger.getLength();
+        WriteLedger currentLedger = getWriteLedger();
+        Map<Long, Long> lastAddsConfirmed = new HashMap<>();
+        boolean anythingChanged = false;
+        for (Write w : writes) {
+            if (w.isDone() || !w.getWriteLedger().ledger.isClosed()) {
+                continue;
+            }
 
-        // Get the writes to execute from the queue.
-        return this.writes.getWritesToExecute(maxTotalSize);
+            // Write likely failed because of LedgerClosedException. Need to check the LastAddConfirmed for each
+            // involved Ledger and see if the write actually made it through or not.
+            long lac = tryGetLastAddConfirmed(w.getWriteLedger(), lastAddsConfirmed);
+            if (w.getEntryId() >= 0 && w.getEntryId() <= lac) {
+                // Write was actually successful. Complete it and move on.
+                w.complete();
+                anythingChanged = true;
+            } else if (currentLedger.ledger.getId() != w.getWriteLedger().ledger.getId()) {
+                // Current ledger has changed; attempt to write to the new one.
+                w.setWriteLedger(currentLedger);
```

Comment:
Here (in this line) you are in `handleClosedLedger`. This is invoked by the `processPendingWrites` which is invoked by the `writeProcessor` (you need to open the whole file for context - check out the constructor). The `writeProcessor` is invoked by:
- A new write being queued
- A write completing (success or failure)
- Rollover completing.

To add more context:
- handleClosedLedgers is invoked just before attempting to write - so right after this method exits the writes are being sent out
- WriteProcessor is built in such a way that regardless of how many requests to have it run, there is only one active execution at any given time (they don't overlap). If multiple requests come in while it's executing, it only re-executes once (not how many times it got requested) - that's because a single execution takes care of everything, so more than one is just wasting CPU.